### PR TITLE
Jk/cumulus 3806 2

### PR DESCRIPTION
**Pull request by Jkovarik**
**Summary:** 

Addresses [CUMULUS-3806: Develop amazing new feature](https://bugs.earthdata.nasa.gov/browse/CUMULUS-3806)

This is PR 2  for this effort: 

- **CUMULUS-3806**
  - Update `@cumulus/db/search` to allow for ordered collation as a
    dbQueryParameter
  - Update `@cumulus/db/search` to allow `dbQueryParameters.limit` to be set to
    `null` to allow for optional unlimited page sizes in search results
  - Update/add type annotations/logic fixes to `@cumulus/api` reconciliation report code
  - Annotation/typing fixes to `@cumulus/cmr-client`
  - Typing fixes to `@cumulus/db`
  - Re-enable Reconciliation Report integration tests
  - Update `@cumulus/client/CMR.getToken` to throw if a non-launchpad token is requested without a username
  - Update `Inventory` and `Granule Not Found` reports to query postgreSQL
    database instead of elasticsearch
  - Update `@cumulus/db/lib/granule.getGranulesByApiPropertiesQuery` to
    allow order by collation to be optionally specified
  - Update `@cumulus/db/lib/granule.getGranulesByApiPropertiesQuery` to
    be parameterized and include a modifier on `temporalBoundByCreatedAt`

## PR Checklist

- [x] Update CHANGELOG
- [x] Unit tests
- [ ] Ad-hoc testing - Deploy changes and test manually
- [x] Integration tests

### CODE REVIEW

File: packages/api/lambdas/create-reconciliation-report.js [hunk 1]
```diff
@@ -305,17 +366,21 @@ async function reconciliationReportForCollections(recReportParams) {
...
     // 'Version' as sort_key
     log.debug('Fetching collections from CMR.');
-    const cmrCollectionIds = await fetchCMRCollections(recReportParams);
-    const esCollectionIds = await fetchESCollections(recReportParams);
-    log.info(`Comparing ${cmrCollectionIds.length} CMR collections to ${esCollectionIds.length} Elasticsearch collections`);
+    const cmrCollectionIds = (await fetchCMRCollections(recReportParams)).sort();
+    const dbCollectionIds = (await fetchDbCollections(recReportParams, knex)).sort();
 
-    let nextDbCollectionId = esCollectionIds[0];
+    log.info(`Comparing ${JSON.stringify(cmrCollectionIds)} CMR collections to ${dbCollectionIds} PostgreSQL collections`);
```
Comment (Jkovarik) [inline packages/api/lambdas/create-reconciliation-report.js:372]: Is this a useful log output?

File: packages/api/lambdas/create-reconciliation-report.js [hunk 2]
```diff
@@ -146,31 +226,46 @@ async function fetchCMRCollections({ collectionIds }) {
...
-    const esCollectionItems = await esCollection.aggregateGranuleCollections();
-    esCollectionIds = esCollectionItems.sort();
-  } else {
-    // return all ES collections
-    const esCollection = new ESSearchQueue(esCollectionSearchParams, 'collection', process.env.ES_INDEX);
-    const esCollectionItems = await esCollection.empty();
-    esCollectionIds = esCollectionItems.map(
-      (item) => constructCollectionId(item.name, item.version)
-    ).sort();
+async function fetchDbCollections(recReportParams, knex) {
```
Comment (Jkovarik) [inline packages/api/lambdas/create-reconciliation-report.js:235]: There's probably a reasonable case to be made for this being generalized to the DB package, however both the typing and the specific modifiers pushed it over the fence for me *for now*, we should re-evaluate this when we take on the update/refactor work for recon reports.

File: packages/api/lambdas/create-reconciliation-report.js [hunk 3]
```diff
@@ -32,21 +36,94 @@ const {
...
+  convertToDBCollectionSearchObject,
+  dateStringToDateOrNull,
   initialReportHeader,
 } = require('../lib/reconciliationReport');
 
 const log = new Logger({ sender: '@api/lambdas/create-reconciliation-report' });
 
 const isDataBucket = (bucketConfig) => ['private', 'public', 'protected'].includes(bucketConfig.type);
 
+// Typescript annotations
```
Comment (Jkovarik) [inline packages/api/lambdas/create-reconciliation-report.js:54]: Reviewers:  we could, in theory use a separate types file.    Current convention is annotations in-line for types that aren't shared, this PR introduces a common types.js file that's up for discussion as well.     Feedback on organization welcome.

**Comment by Jkovarik** [comment]
PR No. 1 merged, and this branch has been merged with feature.

File: packages/api/lib/types.js [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+ * @property {string} [startTimestamp] - An optional start timestamp for the report.
+ * @property {string} [endTimestamp] - An optional end timestamp for the report.
+ * @property {string} [reportType] - An optional type of the report.
+ * @property {string} [location]
+ * @property {string} stackName
+ * @property {string} systemBucket
+ * @property {string} [status] - Optional granule status filter for report
+ */
+
+/**
```
Comment (etcart) [inline packages/api/lib/types.js:15]: is there no way to simplify this? (unless I'm missing something) this "enhancedParams" is used only to define EnhancedNormalizedRecReportParams. is there a reason not to define EnhancedNormalziedRecReportParams as a child of NormalizedRecReportParams, and maybe NormalizedRecReportParams as a child of recReportParams? this feels like a more logical and understandable structure to me

File: packages/cmr-client/src/CMR.ts [hunk 5]
```diff
@@ -131,6 +122,12 @@ export class CMR {
    * @returns {Promise.<string | undefined>} the token
    */
   async getToken(): Promise<string | undefined> {
+    if (this.oauthProvider === 'launchpad') {
+      return this.token;
+    }
+    if (!this.username) {
+      throw new Error('Username not specified for non-launchpad CMR client');
+    }
```
Comment (etcart) [inline packages/cmr-client/src/CMR.ts:130]: is this a fix for a problem introduced in this PR? or is this just a bug fix while you're making changes?

File: packages/db/src/search/queries.ts [hunk 6]
```diff
@@ -243,9 +243,10 @@ export const convertQueryStringToDbQueryParameters = (
 
   const dbQueryParameters: DbQueryParameters = {};
   dbQueryParameters.page = Number.parseInt(page ?? '1', 10);
-  dbQueryParameters.limit = Number.parseInt(limit ?? '10', 10);
-  dbQueryParameters.offset = (dbQueryParameters.page - 1) * dbQueryParameters.limit;
-
+  if (limit !== null) {
```
Comment (etcart) [inline packages/db/src/search/queries.ts:246]: is it intended that this be:
null = unlimited
undefined = 10
? having the different null vs undefined behavior seems weird if there isn't a direct desire for it

**Review by etcart — COMMENTED** [review]
I don't think I'm clear enough on the details here to approve, although I have some questions/comments and could be walked through the logic details.

File: packages/api/lib/types.js (hunk 4 above)
Comment (Jkovarik) [reply packages/api/lib/types.js:15]: @etcart  100% agree, but...the union type worked with tscheck/jsdoc annotations whereas using typescript inheritance did not and sent me down more of a rabbit hole than that refactor felt worth at the time.   I'd love an example of this working, and agree enough I'll spend a bit more time on it.

File: packages/cmr-client/src/CMR.ts (hunk 5 above)
Comment (Jkovarik) [reply packages/cmr-client/src/CMR.ts:130]: Sort-of.     The fix to the class typing above was directly related to changes in the recon report (adding typing)  and that change cascaded into needing to handle this case correctly here.

File: packages/db/src/search/queries.ts (hunk 6 above)
Comment (Jkovarik) [reply packages/db/src/search/queries.ts:246]: That's more or less the case - we want a default if it's not defined, null is explicitly setting 'no limit', and feels better as an overload than '0' or the like.  We could also add another queryParam instead if that feels bad.

File: packages/cmr-client/src/CMRSearchConceptQueue.ts [hunk 7]
```diff
@@ -84,23 +78,28 @@ export class CMRSearchConceptQueue {
...
-   *
-   * @returns {Promise<Object>} an item from the CMR search
    */
-  async shift(): Promise<unknown> {
+  async shift(): Promise<T | null> {
     if (this.items.length === 0) await this.fetchItems();
-    return this.items.shift();
+    const item = this.items.shift();
+    // eslint-disable-next-line lodash/prefer-is-nil
+    if (item === null || item === undefined) {
```
Comment (npauzenga) [inline packages/cmr-client/src/CMRSearchConceptQueue.ts:99]: Is there a reason not to use isNil? Just avoiding an import?

File: packages/db/src/lib/collection.ts [hunk 8]
```diff
@@ -27,3 +29,56 @@ export const getCollectionsByGranuleIds = async (
     .groupBy(`${collectionsTable}.cumulus_id`);
   return await RetryOnDbConnectionTerminateError(query, {}, log);
 };
+
+// TODO - This function is going to be super-non-performant
+// We need to identify the specific need here and see if we can optimize
```
Comment (npauzenga) [inline packages/db/src/lib/collection.ts:34]: Would it help to add some logging here so that if users encounter a bottleneck, we know more about the use case?

File: packages/db/src/search/BaseSearch.ts [hunk 9]
```diff
@@ -458,7 +458,13 @@ abstract class BaseSearch {
     const { sort } = dbQueryParameters || this.dbQueryParameters;
     sort?.forEach((key) => {
       if (key.column.startsWith('error')) {
-        searchQuery.orderByRaw(`${this.tableName}.error ->> 'Error' ${key.order}`);
+        searchQuery.orderByRaw(
+          `${this.tableName}.error ->> 'Error' ${key.order}`
+        );
+      } else if (dbQueryParameters?.collate) {
+        searchQuery.orderByRaw(
+          `${key} collate \"${dbQueryParameters.collate}\"`
```
Comment (npauzenga) [inline packages/db/src/search/BaseSearch.ts:466]: Is there enough justification to use `collate` on the base model as opposed to overwriting the child class' buildSortQuery for just e.g. recon reports?

File: packages/db/src/search/queries.ts (hunk 6 above)
Comment (npauzenga) [reply packages/db/src/search/queries.ts:246]: I _think_ that makes sense. If you don't include `limit`, the default is 10. If you explicitly say limit = null, the limit is removed. It is an interesting thing. I think it makes sense.

File: packages/db/tests/lib/test-granule.js [hunk 10]
```diff
@@ -728,6 +728,53 @@ test.serial('getGranulesByApiPropertiesQuery returns correct granules by provide
   );
 });
 
+test.serial('getGranulesByApiPropertiesQuery returns results POSIX/ASCII sorted when collition is set to "C"', async (t) => {
```
Comment (npauzenga) [inline packages/db/tests/lib/test-granule.js:731]: Was the collate behavior present in the previous ES implementation?

**Review by npauzenga — CHANGES_REQUESTED** [review]
Looks great @Jkovarik! The comments above are mostly clarifying questions.

File: packages/api/lambdas/create-reconciliation-report.js (hunk 3 above)
Comment (Jkovarik) [reply packages/api/lambdas/create-reconciliation-report.js:52]: Ok, fair enough!  I'll get a commit here shortly.

Comment (Jkovarik) [reply packages/api/lambdas/create-reconciliation-report.js:52]: Take a look at: b7bd336235a

File: CHANGELOG.md [hunk 11]
```diff
@@ -40,18 +41,29 @@ aws lambda invoke --function-name $PREFIX-ReconciliationReportMigration $OUTFILE
   - Update `@cumlus/api/ecs/async-operation` to not update Elasticsearch index when
     reporting status of async operation
 - **CUMULUS-3806**
+  - Update `@cumulus/db/search` to allow for ordered collation as a
+    dbQueryParameter
+  - Update `@cumulus/db/search` to allow `dbQueryParameters.limit` to be set to
+    `null` to allow for optional unlimited page sizes in search results
```
[the remark this answers is not in the document]
Comment (Jkovarik) [reply CHANGELOG.md:47]: Ordered collation was required due to ordering being of primary importance for the recon reports and the database ordering being inconsistent between ES, the unit environment and production.

File: CHANGELOG.md [hunk 12]
```diff
@@ -21,6 +21,7 @@ aws lambda invoke --function-name $PREFIX-ReconciliationReportMigration $OUTFILE
 
 ### Replace ElasticSearch Phase 2
 
+
```
[the remark this answers is not in the document]
Comment (Jkovarik) [reply CHANGELOG.md]: 74b97a36d6a

File: packages/api/lambdas/create-reconciliation-report.js [hunk 13]
```diff
@@ -101,41 +177,32 @@ function isOneWayGranuleReport(reportParams) {
...
-
-/**
- * fetch CMR collections and filter the returned UMM CMR collections by the desired collectionIds
+ * Fetches collections from the CMR (Common Metadata Repository) and returns their IDs.
  *
- * @param {Object} recReportParams - input report params
- * @param {Array<string>} recReportParams.collectionIds - array of collectionIds to keep
- * @returns {Array<string>} filtered list of collectionIds returned from CMR
+ * @param {EnhancedNormalizedRecReportParams} recReportParams - The parameters for the function.
+ * @returns {Promise<string[]>} A promise that resolves to an array of collection IDs from the CMR.
```
[the remark this answers is not in the document]
Comment (Jkovarik) [reply packages/api/lambdas/create-reconciliation-report.js:137]: I don't think there's a functional difference between `string` and `collectionId` with respect to typing - you can't enforce string patters as a matter of >type<.

File: packages/api/lambdas/create-reconciliation-report.js [hunk 14]
```diff
@@ -145,31 +212,42 @@ async function fetchCMRCollections({ collectionIds }) {
...
+      constructCollectionId(collection.name, collection.version));
   }
-  return esCollectionIds;
+
+  const queryStringParameters = removeNilProperties({
+    _id__in: collectionIds ? collectionIds.join(',') : undefined,
+    timestamp__from: startTimestamp,
+    timestamp__to: endTimestamp,
+    sort_key: ['name', 'version'],
+    collate: 'C',
```
[the remark this answers is not in the document]
Comment (Jkovarik) [reply packages/api/lambdas/create-reconciliation-report.js:197]: It's referring to `COLLATE` in postgres, which allows specific ordering in query or table level operations: 

https://www.postgresql.org/docs/current/collation.html

File: packages/api/lambdas/create-reconciliation-report.js [hunk 15]
```diff
@@ -192,7 +270,7 @@ async function createReconciliationReportForBucket(Bucket, recReportParams) {
     sortColumns: ['key'],
     granuleColumns: ['granule_id'],
     collectionIds: recReportParams.collectionIds,
-    providers: recReportParams.providers,
+    providers: recReportParams.provider,
```
[the remark this answers is not in the document]
Comment (Jkovarik) [reply packages/api/lambdas/create-reconciliation-report.js]: Yikes - the better ask is 'why isn't this typed' - you're entirely right, this is a bad change. 

Fixed, and added the typing as I'm already there: 

7b8e9dd896b

File: packages/api/lambdas/create-reconciliation-report.js [hunk 16]
```diff
@@ -464,14 +549,17 @@ exports.reconciliationReportForGranuleFiles = reconciliationReportForGranuleFile
...
- *                                                 (e.g. { bucket: distribution path })
- * @param {Object} params.recReportParams        - Lambda report paramaters for narrowing focus
- * @returns {Promise<Object>}                    - an object with the granulesReport and filesReport
+ * @param {Object} params                          - parameters
+ * @param {string} params.collectionId             - the collection which has the granules to be
+ *                                                   reconciled
+ * @param {Object} params.bucketsConfig            - bucket configuration object
+ * @param {Object} params.distributionBucketMap    - mapping of bucket->distirubtion path values
+ *                                                   (e.g. { bucket: distribution path })
+ * @param {EnhancedNormalizedRecReportParams} params.recReportParams - Lambda report paramaters for
```
[the remark this answers is not in the document]
Comment (Jkovarik) [reply packages/api/lambdas/create-reconciliation-report.js:512]: Naming *is* among the most difficult problems in computer science.    I'm open to suggestions, of course :)

File: packages/api/lambdas/create-reconciliation-report.js [hunk 17]
```diff
@@ -538,18 +641,24 @@ async function reconciliationReportForGranules(params) {
       } else {
         // Found an item that is in both CMR and Cumulus database
         granulesReport.okCount += 1;
+        // eslint-disable-next-line no-await-in-loop
+        const postgresGranuleFiles = await getFilesAndGranuleInfoQuery({
+          knex,
+          searchParams: { granule_cumulus_id: nextDbItem.cumulus_id },
+          sortColumns: ['key'],
```
[the remark this answers is not in the document]
Comment (Jkovarik) [reply packages/api/lambdas/create-reconciliation-report.js:602]: I believe so, yes, in this case, though, minor order differences probably aren't meaningful.

File: packages/api/lambdas/create-reconciliation-report.js [hunk 18]
```diff
@@ -578,6 +690,9 @@ async function reconciliationReportForGranules(params) {
     if (!oneWay) {
       while (await cmrGranulesIterator.peek()) { // eslint-disable-line no-await-in-loop
         const cmrItem = await cmrGranulesIterator.shift(); // eslint-disable-line no-await-in-loop
+        if (!cmrItem) {
+          throw new Error('CMR returned item is null in reconciliationReportForGranules');
+        }
```
[the remark this answers is not in the document]
Comment (Jkovarik) [reply packages/api/lambdas/create-reconciliation-report.js:649]: Same as above - this is as much about properly typing/handling different types instead of relying on lower methods that don't have that purpose.

File: packages/api/lambdas/create-reconciliation-report.js [hunk 19]
```diff
@@ -607,20 +722,21 @@ exports.reconciliationReportForGranules = reconciliationReportForGranules;
...
  *                                                 (e.g. { bucket: distribution path })
- * @param {Object} [params.recReportParams]      - optional Lambda endpoint's input params to
- *                                                 narrow report focus
- * @param {number} [params.recReportParams.StartTimestamp]
- * @param {number} [params.recReportParams.EndTimestamp]
- * @param {string} [params.recReportparams.collectionIds]
- * @returns {Promise<Object>}                    - a reconciliation report
+ * @param {EnhancedNormalizedRecReportParams} params.recReportParams - Lambda endpoint's input
+ * params to narrow focus of report
+ * @returns {Promise<Object>}                      - a reconciliation report
```
[the remark this answers is not in the document]
Comment (Jkovarik) [reply packages/api/lambdas/create-reconciliation-report.js:685]: The answer to that is almost always yes, but in this case, I felt that exceeded the scope of this ticket - nothing I'm doing directly changes the schema of the output object.       That is, of course, a matter of opinion, so feel free to push back on it.

File: packages/api/lambdas/create-reconciliation-report.js [hunk 20]
```diff
@@ -818,7 +928,7 @@ async function processRequest(params) {
     reportName,
     systemBucket,
     stackName,
-    knex = await getKnexClient(env),
+    knex = await getKnexClient({ env }),
```
[the remark this answers is not in the document]
Comment (Jkovarik) [reply packages/api/lambdas/create-reconciliation-report.js:885]: The typing was incorrect, yes.       It worked in that the client happens to correctly use env defaults.

File: packages/api/lambdas/create-reconciliation-report.js [hunk 21]
```diff
@@ -856,18 +967,20 @@ async function processRequest(params) {
       log.error(
         'Internal Reconciliation Reports are no longer valid, as Cumulus is no longer utilizing Elasticsearch'
       );
+      //TODO remove internal rec report code
```
[the remark this answers is not in the document]
Comment (Jkovarik) [reply packages/api/lambdas/create-reconciliation-report.js:924]: No, that's PR #3 for this effort.

File: packages/api/lambdas/create-reconciliation-report.js [hunk 22]
```diff
@@ -856,18 +967,20 @@ async function processRequest(params) {
...
       throw new Error('Internal Reconciliation Reports are no longer valid');
     } else if (reportType === 'Granule Inventory') {
       await createGranuleInventoryReport(recReportParams);
     } else if (reportType === 'ORCA Backup') {
       await createOrcaBackupReconciliationReport(recReportParams);
-    } else {
+    } else if (['Inventory', 'Granule Not Found'].includes(reportType)) {
```
[the remark this answers is not in the document]
Comment (Jkovarik) [reply packages/api/lambdas/create-reconciliation-report.js:930]: It's more explicit with respect to typing - we don't actually want <other> things to be passed in, and the API doesn't accept them.         Having 'foobar' or 'some type I think you supported but you do not' return a report should be unexpected behavior.

File: packages/api/lib/reconciliationReport.js [hunk 23]
```diff
@@ -9,6 +11,32 @@ const Logger = require('@cumulus/logger');
...
+ * @property {string | string[] | undefined} granuleIds - The granule IDs.
+ * @property {string | string[] | undefined} provider - The provider.
+ * @property {string | string[] | undefined} providers - The providers.
+ * @property {string | undefined} location - The location.
+ * @property {string | undefined} reportEndTime - The end time of the report.
+ * @property {string | undefined} reportStartTime - The start time of the report.
+ * @property {string} reportType - The type of the report.
+ * @property {string} status - The status of the report.
+ */
+
```
[the remark this answers is not in the document]
Comment (Jkovarik) [reply packages/api/lib/reconciliationReport.js:21]: 1af2d1cb8f1

File: packages/api/lib/reconciliationReport/normalizeEvent.js [hunk 24]
```diff
@@ -1,15 +1,22 @@
...
+ * @typedef {import('../types').RecReportParams } RecReportParams
+ * @typedef {import('../types').NormalizedRecReportParams } NormalizedRecReportParams
+ */
+
 /**
  * ensures input reportType can be handled by the lambda code.
  *
  * @param {string} reportType
- * @returns {undefined} - if reportType is valid
+ * @returns {void} - if reportType is valid
```
[the remark this answers is not in the document]
Comment (Jkovarik) [reply packages/api/lib/reconciliationReport/normalizeEvent.js:19]: In this case, the method doesn't *return* anything, so `void` is correct.   Undefined is an undefined return object.     It's a typing error.

File: packages/api/lib/reconciliationReport/normalizeEvent.js [hunk 25]
```diff
@@ -78,34 +78,44 @@ function updateCollectionIds(collectionId, modifiedEvent) {
...
  *
- * @param {Object} event - input payload
- * @returns {Object} - Object with normalized parameters
+ * @param {RecReportParams} event - input payload
+ * @returns {NormalizedRecReportParams} - Object with normalized parameters
  */
 function normalizeEvent(event) {
   const systemBucket = event.systemBucket || process.env.system_bucket;
+  if (!systemBucket) {
+    throw new InvalidArgument('systemBucket is required.');
```
[the remark this answers is not in the document]
Comment (Jkovarik) [reply packages/api/lib/reconciliationReport/normalizeEvent.js]: Great catch: 244b5c5f8ae

File: packages/api/tests/lambdas/test-create-reconciliation-report-internals.js [hunk 26]
```diff
@@ -86,39 +85,6 @@ test(
   }
 );
 
-test(
-  'shouldAggregateGranulesForCollections returns true only when one or more specific parameters '
```
[the remark this answers is not in the document]
Comment (Jkovarik) [reply packages/api/tests/lambdas/test-create-reconciliation-report-internals.js:90]: This method was removed/the code was refactored to remove it as part of the functional updates to move from ES to RDS

File: packages/cmr-client/src/CMR.ts [hunk 27]
```diff
@@ -67,29 +67,20 @@ export interface CMRConstructorParams {
   *  clientId: 'my-clientId',
   *  token: 'cmr_or_launchpad_token'
   * });
+  * TODO: this should be subclassed or refactored to a functional style
```
[the remark this answers is not in the document]
Comment (Jkovarik) [reply packages/cmr-client/src/CMR.ts:70]: I'm not sure this is an actionable ticket so much as a maintenance note to whomever touches this next.

File: packages/cmr-client/src/CMRSearchConceptQueue.ts [hunk 28]
```diff
@@ -84,23 +78,28 @@ export class CMRSearchConceptQueue {
    * This does not remove the object from the queue.  When there are no more
    * items in the queue, returns 'null'.
    *
-   * @returns {Promise<Object>} an item from the CMR search
    */
-  async peek(): Promise<unknown> {
+  async peek(): Promise<T> {
     if (this.items.length === 0) await this.fetchItems();
+    if (this.items[0] === null) {
+      return null as unknown as T;
```
[the remark this answers is not in the document]
Comment (Jkovarik) [reply packages/cmr-client/src/CMRSearchConceptQueue.ts]: Actually.... yes, because I'm just going to fix/make it more clear: 

da07f81e856

File: packages/cmr-client/src/CMRSearchConceptQueue.ts (hunk 7 above)
Comment (Jkovarik) [reply packages/cmr-client/src/CMRSearchConceptQueue.ts:99]: Because isNil doesn't work properly with ts-check, in that it doesn't act as a type-guard.      I could work around that in type annotations, but that feels even less maintainable.

File: packages/db/src/lib/collection.ts (hunk 8 above)
Comment (Jkovarik) [reply packages/db/src/lib/collection.ts:34]: I'm not entirely sure what that looks like, as the bottleneck is likely *in* the SQL server logs. 

I feel this probably falls *solidly* out of scope for this ticket, but open to talking about it.

File: packages/db/src/search/BaseSearch.ts (hunk 9 above)
Comment (Jkovarik) [reply packages/db/src/search/BaseSearch.ts:466]: I think so, yes - this is generally useful in the case that sorting order matters with respect to CMR or other external data sources.

File: packages/db/tests/lib/test-granule.js (hunk 10 above)
Comment (Jkovarik) [reply packages/db/tests/lib/test-granule.js:731]: Not explicitly, but implicitly because of the default methods ES uses to sort - that's why the tests all broke due to sort order when uppercase/lowercase granuleIds are involved.

File: packages/api/lib/types.js (hunk 4 above)
Comment (Jkovarik) [reply packages/api/lib/types.js:15]: @etcart  re-reading this, I think I grok what you were getting at - the problem is this sort of thing: 

```
/**
 * @typedef {NormalizedRecReportParams} EnhancedNormalizedRecReportParams
```

Winds up not properly representing the combined type.   I'm uncertain if there's a better way to handle this.

Comment (Jkovarik) [reply packages/api/lib/types.js:15]: (talked about this in a huddle, closing with @etcart 's approval)

File: packages/api/lambdas/create-reconciliation-report.js (hunk 3 above)
Comment (npauzenga) [reply packages/api/lambdas/create-reconciliation-report.js:52]: Nice. Thanks!

File: CHANGELOG.md (hunk 12 above)
[the remark this answers is not in the document]
Comment (npauzenga) [reply CHANGELOG.md]: Critical change. Thanks.

File: packages/api/lambdas/create-reconciliation-report.js (hunk 13 above)
[the remark this answers is not in the document]
Comment (npauzenga) [reply packages/api/lambdas/create-reconciliation-report.js:136]: Yeah, fair enough. Any new type would be basically just an alias to `string[]`, which is gross.

File: packages/api/lambdas/create-reconciliation-report.js (hunk 15 above)
[the remark this answers is not in the document]
Comment (npauzenga) [reply packages/api/lambdas/create-reconciliation-report.js]: So you're saying typing is good? 😄

File: packages/api/lambdas/create-reconciliation-report.js (hunk 16 above)
[the remark this answers is not in the document]
Comment (npauzenga) [reply packages/api/lambdas/create-reconciliation-report.js:511]: Nah. You can jump to the type def so it's not a big deal. I pause on things like this wondering if there's value in types that are... not super clear in what they mean. I feel like (because this is only vibes) the primitive types spelled out in the docstring are more immediately legible but the difference is unbelievably small. And if we're ever using the same params object in more than one place, it's valuable to abstract.

tl;dr it's fine. Just wanted to chat 😆

File: packages/api/lambdas/create-reconciliation-report.js (hunk 17 above)
[the remark this answers is not in the document]
Comment (npauzenga) [reply packages/api/lambdas/create-reconciliation-report.js:601]: Meaning sorting by `key` AND `bucket` probably isn't significantly meaningful?

File: packages/api/lambdas/create-reconciliation-report.js (hunk 18 above)
[the remark this answers is not in the document]
Comment (npauzenga) [reply packages/api/lambdas/create-reconciliation-report.js:648]: Yup, nice. The reason for asking these questions is mostly "Is this is a new way to fail?". The answer is 'nope' so we're good.

File: packages/api/lambdas/create-reconciliation-report.js (hunk 19 above)
[the remark this answers is not in the document]
Comment (npauzenga) [reply packages/api/lambdas/create-reconciliation-report.js:684]: Nah, primitives are fine. We probably wanna draw the line at "no typing" or `any`.

File: packages/api/lambdas/create-reconciliation-report.js [hunk 29]
```diff
@@ -668,7 +784,7 @@ async function reconciliationReportForCumulusCMR(params) {
  * @param {Object} report       - report to upload
  * @param {string} systemBucket - system bucket
  * @param {string} reportKey    - report key
- * @returns {Promise<null>}
+ * @returns {Promise}
```
[the remark this answers is not in the document]
Comment (npauzenga) [reply packages/api/lambdas/create-reconciliation-report.js]: @Jkovarik I am curious what you think about this. Not particularly for this use but generally. What's your opinion on importing aws types and using them in our docstrings for e.g. return type declarations?

File: packages/api/lambdas/create-reconciliation-report.js (hunk 20 above)
[the remark this answers is not in the document]
Comment (npauzenga) [reply packages/api/lambdas/create-reconciliation-report.js:884]: Nice. good catch.

File: packages/api/lambdas/create-reconciliation-report.js (hunk 29 above)
[the remark this answers is not in the document]
Comment (Jkovarik) [reply packages/api/lambdas/create-reconciliation-report.js]: I'm fine with being explicit when it's needed, but if it involves adding a dependency I'm also fine with the resolved type, so long as that works with the team's tooling.

File: packages/db/src/lib/collection.ts (hunk 8 above)
Comment (npauzenga) [reply packages/db/src/lib/collection.ts:34]: Nope, only asking in case we could quickly add logs here.

File: packages/api/lambdas/create-reconciliation-report.js (hunk 17 above)
[the remark this answers is not in the document]
Comment (Jkovarik) [reply packages/api/lambdas/create-reconciliation-report.js:601]: I'm fairly sure the original sort in ES was key only, so it maps to what was being used before.    

When I say 'not meaningful', I'm mostly suggesting that the ordering of the files probably isn't actionable (we're not promising return order, and not relying on it).

File: packages/api/lambdas/create-reconciliation-report.js [hunk 30]
```diff
@@ -766,7 +874,7 @@ async function createReconciliationReport(recReportParams) {
               + bucketGranuleCount;
           });
         } else {
-          delete report.filesInCumulus.okCountByGranule;
+          report.filesInCumulus.okCountByGranule = {};
```
[the remark this answers is not in the document]
Comment (npauzenga) [reply packages/api/lambdas/create-reconciliation-report.js:830]: Don't we still need to specify the imported type? ` @returns {PutObjectCommandOutput} - ...`?

[the remark this answers is not in the document]
Comment (Jkovarik) [reply packages/api/lambdas/create-reconciliation-report.js:830]: So - the problem is, to import an aws type like that we have to *import* like that commit matches, and to properly do that we'd have to put a direct dependency on aws-sdk which is *not* currently a direct dependency. 

Probably the proper way to do this explicitly is to export the type from the cumulus/aws-client package, but the approach here utilizes ts-check/typescript's inferred return type and displays the type properly in the tooling (but obviously not in the code).      

The typing is consistent either way (ts-check/lint know it's of the type PutObjectCommandOutput without needing to import it, but it does, *in theory* make the tooling required to know the return type.

[the remark this answers is not in the document]
Comment (Jkovarik) [reply packages/api/lambdas/create-reconciliation-report.js:830]: See this response as well https://github.com/nasa/cumulus/pull/3811#discussion_r1808982868

[the remark this answers is not in the document]
Comment (npauzenga) [reply packages/api/lambdas/create-reconciliation-report.js:830]: Yeah, it definitely would require a new dependency. Not sure it's worth the trouble.

**Review by npauzenga — APPROVED** [review]
Changes look great @Jkovarik. Thanks for addressing my comments!

**Comment by Jkovarik** [comment]
Thanks for the review @npauzenga  🙇🏻
